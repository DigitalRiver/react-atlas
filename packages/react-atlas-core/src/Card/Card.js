import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";

/**
 * Simple Card component that wraps a div around content with card styling.
 */
class Card extends React.PureComponent {
  constructor(props) {
    super(props);

  }
  render() {
    const { children, className, legend, style, standardFieldset, image, title, width } = this.props;
    let Width = width ? width + "px" : null;
    let cardBody = children ?
      <div
        styleName={"cardBody"}
      >
        {children}
      </div>
      : null;

    let isFieldset = standardFieldset ? 
      <fieldset style={{ "width": Width}} styleName={"card"} className={cx(className)}>
        {legend && <legend styleName={"legend"}>{legend}</legend>}
        {children}
      </fieldset>
      : null;

    let useImage = image ? 
      <img
        src={image}
        styleName={"image"}
      />
      : null;

    let Title = title ?
      <div styleName={"title"}>{title}</div>
      : null;

    let card = !standardFieldset ? 
      <div        
        style={{ "width": Width}}
        styleName={"card"} 
      >
        {useImage}
        {Title}
        {cardBody}
      </div>
      : null;

    return (
      <div 
        style={style}
        className={cx(className)}
      >
        {isFieldset}
        {card}
      </div>
    );
  }
}

Card.propTypes = {
  /**
   * Any HTML element or React Component.
   * @examples <p>Some Text.</p>
   */
  "children": PropTypes.node.isRequired,

  /** An Object, array, or string of CSS classes to apply to card.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /**
   * Title of the Card
   */
  "legend": PropTypes.string,

  /**
   * Pass inline styles here.
   */
  "style": PropTypes.object,

  /**
   * Path to an image
   * @examples "http://path.to/an/image.jpg"
   */
   "image": PropTypes.string,

  /**
   * A string. 
   * @examples "Title Words" will output "Title Words"
   */
  "title": PropTypes.string,

  /**
   * A width. 
   */
  "width": PropTypes.number,

  /**
   * Used for non fieldset cards
   * @example <Card standardFieldset>{children}</Card>
   */
  "standardFieldset": PropTypes.boolean
};

export default Card;