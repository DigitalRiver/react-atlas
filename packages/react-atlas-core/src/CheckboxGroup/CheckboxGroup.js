import React, { cloneElement } from "react";
import PropTypes from "prop-types";
import cx from "classnames";

class CheckboxGroup extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      "totalChecked": this.initialChecked(),
      "groupError": false
    };
  }

  // Loop through child components to determine how many <Checkbox /> components are checked by default
  initialChecked() {
    let newChecked = 0;
    React.Children.map(this.props.children, child => {
      if (child.props.checked) {
        newChecked++;
      }
    });
    return newChecked;
  }

  // When child <Checkbox /> component changes state, update "totalChecked" value and verify against min or max props value
  handleChange = (value, event, isValid, checked) => {
    event.persist();
    const newChecked = checked
      ? this.state.totalChecked + 1
      : this.state.totalChecked - 1;
    if (checked) {
      this.setState({ "totalChecked": newChecked });
    } else {
      this.setState({ "totalChecked": newChecked });
    }
    if (
      this.props.max && newChecked > this.props.max ||
      this.props.min && newChecked < this.props.min
    ) {
      this.setState({ "groupError": true });
    } else {
      this.setState({ "groupError": false });
    }

    if (typeof this.props.onChange !== "undefined") {
      this.props.onChange(value, event, isValid, checked);
    }
  };

  // If min or max props are set and the correct number of checkboxes aren't checked, display the appropriate error message
  maxMinMessage = () => {
    let fullMessage;
    if (this.props.max && this.state.totalChecked > this.props.max) {
      if (this.props.limitMessage) {
        fullMessage = this.props.limitMessage.replace("{0}", this.props.max);
        return fullMessage;
      } else {
        return `Please select no more than ${this.props
          .max} of the options below`;
      }
    } else if (this.props.min && this.state.totalChecked < this.props.min) {
      if (this.props.limitMessage) {
        fullMessage = this.props.limitMessage.replace("{0}", this.props.min);
        return fullMessage;
      } else {
        return (
          "Please select at least " + this.props.min + " of the options below"
        );
      }
    } else {
      return false;
    }
  };

  render() {
    const {
      className,
      children,
      name,
      inline,
      inlineChildren,
      title,
      style
    } = this.props;
    const checkboxGroupStyles = cx({
      "checkboxGroup": !inline,
      "inline": inline
    });
    return (
      <div
        style={style}
        className={cx(className)}
        styleName={checkboxGroupStyles}
      >
        {title &&
          <div styleName={"header"}>
            <span styleName={"headerFont"}>{title}</span>
            {this.state.groupError &&
              <span styleName={"error_message"}>{this.maxMinMessage()}</span>
            }
          </div>
        }
        {React.Children.map(children, child => {
          child = cloneElement(child, {
            "inline": inlineChildren,
            "name": name,
            "onChange": this.handleChange,
            "groupError": this.state.groupError
          });
          return child;
        })}
      </div>
    );
  }
}

CheckboxGroup.propTypes = {
  /**
   * An optional callback that is fired when any checkbox in the
   * CheckboxGroup changes value. The callback has four arguments.
   * function(value, event, isValid, checked).
   */
  "onChange": PropTypes.func,
  /**
   * Anything that can be in a checkbox group. Typically only includes Checkbox components and a header.
   * @examples '<CheckboxGroup><Checkbox/><Checkbox/></CheckboxGroup>'
   */
  "children": PropTypes.node.isRequired,
  /** An Object, array, or string of CSS classes to apply to CheckboxGroup.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
   * Form name for the element, this will set all Checkbox children the same form name.
   * @examples '<CheckboxGroup name="test"></CheckboxGroup>'
   */
  "name": PropTypes.string,
  /**
   * Displayed header text for the CheckboxGroup.
   * @examples '<CheckboxGroup title="Options"></CheckboxGroup>'
   */
  "title": PropTypes.string,
  /**
   * Defines if the checkbox group should display as an inline element.
   * @examples '<CheckboxGroup inline></CheckboxGroup>'
   */
  "inline": PropTypes.bool,
  /**
   * Defines if the checkbox group should display child Checkbox components inline.
   * @examples '<CheckboxGroup inlineChildren></CheckboxGroup>'
   */
  "inlineChildren": PropTypes.bool,
  /**
   * Defines a minimum number of checkboxes the user must select in order to proceed.
   * @examples '<CheckboxGroup min="3"></CheckboxGroup>'
   */
  "min": PropTypes.number,
  /**
   * Defines a maximum number of checkboxes the user is allowed to select before proceeding.
   * @examples '<CheckboxGroup max="3"></CheckboxGroup>'
   */
  "max": PropTypes.number,
  /**
   * Defines a custom message to show if either min or max properties are not met. {0} can be used in the message to be replaced by the appropriate min or max property.
   * @examples '<CheckboxGroup min="3" limitMessage="Please select at least {0} of the checkboxes below."></CheckboxGroup>'
   */
  "limitMessage": PropTypes.string,

  /**
   * Pass inline styles here.
   */
  "style": PropTypes.object
};

CheckboxGroup.defaultProps = {
  "inline": false
};

export default CheckboxGroup;
