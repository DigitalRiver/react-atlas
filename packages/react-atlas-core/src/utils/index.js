import events from "./events";
import prefixer from "./prefixer";
// import time from './time';
import utils from "./utils";
import classNames from "./classNames";

export { events, prefixer, utils, classNames };
