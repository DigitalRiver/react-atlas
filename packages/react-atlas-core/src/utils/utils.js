export default {
  angleFromPositions(cx, cy, ex, ey) {
    const theta = Math.atan2(ey - cy, ex - cx) + Math.PI / 2;
    return theta * 180 / Math.PI;
  },
  angle360FromPositions(cx, cy, ex, ey) {
    const angle = this.angleFromPositions(cx, cy, ex, ey);
    return angle < 0 ? 360 + angle : angle;
  },
  range(start = 0, stop = null, step = 1) {
    let [_start, _stop] = [0, start];
    if (stop !== null) {
      [_start, _stop] = [start, stop];
    }
    const length = Math.max(Math.ceil((_stop - _start) / step), 0);
    const range = Array(length);

    for (let idx = 0; idx < length; idx++, _start += step) {
      range[idx] = _start;
    }

    return range;
  },
  round(number, decimals) {
    if (!isNaN(parseFloat(number)) && isFinite(number)) {
      const decimalPower = Math.pow(10, decimals);
      return Math.round(parseFloat(number) * decimalPower) / decimalPower;
    }
    return NaN;
  },
  getViewport() {
    return {
      "height": window.innerHeight || document.documentElement.offsetHeight,
      "width": window.innerWidth || document.documentElement.offsetWidth
    };
  },
  cloneObject(object) {
    return JSON.parse(JSON.stringify(object));
  },
  inputTypeForPrototype(prototype) {
    if (prototype === Date) {
      return "date";
    }
    if (prototype === Number) {
      return "number";
    }
    if (prototype === Boolean) {
      return "checkbox";
    }
    return "text";
  },
  prepareValueForInput(value, type) {
    if (type === "date") {
      return new Date(value).toISOString().slice(0, 10);
    }
    if (type === "checkbox") {
      return value ? "on" : null;
    }
    return value;
  },
  getSelection(elem) {
    let start, end, rangeElem, clone;

    if (elem.selectionStart !== undefined) {
      start = elem.selectionStart;
      end = elem.selectionEnd;
    }
    else {
      try {
        elem.focus();
        rangeElem = elem.createTextRange();
        clone = rangeElem.duplicate();

        rangeElem.moveToBookmark(document.selection.createRange().getBookmark());
        clone.setEndPoint('EndToStart', rangeElem);

        start = clone.text.length;
        end = start + rangeElem.text.length;
      }
      catch (e) { /* not focused or not visible */ }
    }

    return { start, end };
  },
  setSelection(elem, selection) {
    let rangeElem;

    try {
      if (elem.selectionStart !== undefined) {
        elem.focus();
        elem.setSelectionRange(selection.start, selection.end);
      }
      else {
        elem.focus();
        rangeElem = elem.createTextRange();
        rangeElem.collapse(true);
        rangeElem.moveStart('character', selection.start);
        rangeElem.moveEnd('character', selection.end - selection.start);
        rangeElem.select();
      }
    }
    catch (e) { /* not focused or not visible */ }
  }
};
