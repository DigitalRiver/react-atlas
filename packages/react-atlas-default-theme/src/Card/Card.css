.card {
	composes: marg-b-2 from '../styles.css';
	composes: rounded from '../styles.css';
	composes: bg-white from '../styles.css';
	composes: default-border from '../styles.css';
	composes: default-text from '../styles.css';
}

.legend {
	composes: header-font from '../styles.css';
}

.image {
  display: block;
  height: 100%;
  width: 100%;
  background-position: center;
  background-size: cover;
}

.title {
	composes: header-font from '../styles.css';
	composes: secondary-text from '../styles.css';
	composes: marg-t-2 from '../styles.css';
	composes: pad-h-2 from '../styles.css';
	word-wrap: break-word;
}

.cardBody {
	composes: default-font from '../styles.css';
	composes: default-text from '../styles.css';
	composes: pad-h-2 from '../styles.css';
	composes: marg-b-2 from '../styles.css';
	word-wrap: break-word;
}

