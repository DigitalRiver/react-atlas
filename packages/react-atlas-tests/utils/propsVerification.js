export function verifyPropsDefaultValue(component, expectedProps) {
  //Compares each property value of the component with the expected value

  var result = true;
  var partialResult;

  for (var [compPropName, compPropValue] of expectedProps.entries()) {
    partialResult = component.props()[compPropName] == compPropValue;

    if (!partialResult) {
      console.log(
        "[" +
          compPropName +
          "] -> Got: " +
          component.props()[compPropName] +
          " - expected: " +
          compPropValue
      );
      result = false;
    }
  }
  return result;
}
