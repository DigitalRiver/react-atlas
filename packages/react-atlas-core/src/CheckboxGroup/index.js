import { default as CheckboxGroupCore } from "./CheckboxGroup";
export { CheckboxGroupCore };