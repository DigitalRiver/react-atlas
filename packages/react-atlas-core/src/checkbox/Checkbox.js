import React, { Component, PropTypes } from 'react';
import themeable from 'react-themeable';
import { classNames } from '../utils/utils';


/**
 * Simple component for a basic checkbox
 */
const Checkbox = ({title, label, disabled, inline, ...other}) => {
      //const cx = ClassNames.bind(other.theme);
      const theme = themeable(other.theme);
      const componentClasses = classNames({
          "block": !inline,
          inline,
          disabled,
          [className]: className
      });

      title = title ? title : label;

      return (
          <label {...theme(1, ...componentClasses)} title={title}>
              <input {...other} type="checkbox" disabled={disabled} {...theme(2, 'input')} />
              {label && <span {...theme(3, 'label')}>{label}</span>}
          </label>
      );
};

Checkbox.propTypes = {
  /**
   * Text for checkbox label
   * @examples 'Some Label'
   */
  label: PropTypes.string,
	/**
   * A css class name that will be appended to the wrapping <label> element around the <input> and <span> elements.
   */
  className: PropTypes.string,
	/**
   * If included, checkbox is disabled
   * @examples <Checkbox disabled />, <Checkbox disabled={true} />
   */
  disabled: PropTypes.bool,
	/**
   * If included, renders the checkbox and it's label inline, so it can be side-by-side to other content.
   */
  inline: PropTypes.bool,
  /**
   * Text for checkbox label title. (i.e. "alt-text" for checkboxes, useful for accessibility). If not provided, will be label text.
   * @examples 'Some Title'
   */
  title: PropTypes.string,
	/**
   * Defines if checkbox should be checked on load.
   */
  defaultChecked: PropTypes.bool
};

Checkbox.defaultProps = {
  className: '',
  disabled: false,
  inline: false
};

Checkbox.styleguide = {
  category: "Form Components",
  index: "3.2",
  example: `
<section style={{textAlign:"left"}}>
  <h4 style={{marginBottom: "10px"}}>Checkboxes</h4>
  <Checkbox
    defaultChecked={true}
    label="Checked"
  />
  <Checkbox
    defaultChecked={false}
    label="Not checked"

  />
  <Checkbox
    defaultChecked={true}
    label="Disabled checkbox"
    disabled
  />
  <Checkbox
    defaultChecked={false}
    label="Inline checkbox 1"
    inline
  />
  <Checkbox
    defaultChecked={false}
    label="Inline checkbox 2"
    inline
  />
</section>
`
};

export default Checkbox;
