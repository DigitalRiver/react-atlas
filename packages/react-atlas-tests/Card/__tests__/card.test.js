import React from "react";
import { mount } from "enzyme";
import { CardCore } from "../../../react-atlas-core/src/Card/index";
import { default as Checkbox } from "../../../react-atlas-core/src/Checkbox/Checkbox";
import { default as CheckboxGroup } from "../../../react-atlas-core/src/CheckboxGroup/CheckboxGroup";
import { verifyPropsDefaultValue } from "../../utils/propsVerification";

import renderer from "react-test-renderer";

describe("Test Card component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<CardCore className={"name"} legend={"lalala"} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Card - Basic properties test", function() {
    const result = mount(<CardCore className={"name"} legend={"lalala"} />);
    const expected = new Map([["className", "name"], ["legend", "lalala"]]);
    expect(verifyPropsDefaultValue(result, expected)).toEqual(true);
  });

  it("Card - Children test", function() {
    const result = mount(
      <CardCore legend="Card Example">
        <CheckboxGroup title="Checkbox Group" name="checkboxGroup">
          <Checkbox label="Checkbox 1" checked />
          <Checkbox label="Checkbox 2" />
          <Checkbox label="Checkbox 3" />
        </CheckboxGroup>
      </CardCore>
    );
  });
});
