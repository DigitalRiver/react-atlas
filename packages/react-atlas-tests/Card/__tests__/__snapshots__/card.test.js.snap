// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Card component renders correctly 1`] = `
<fieldset
  className="name"
  style={undefined}
  styleName="card"
>
  <legend
    styleName="legend"
  >
    lalala
  </legend>
  <p>
    Some card text.
  </p>
</fieldset>
`;
